src/Compilers/VisualBasic/Portable/Compilation/VisualBasicDeterministicKeyBuilder.vb

```diff
@@ -0,0 +1,109 @@
+﻿' Licensed to the .NET Foundation under one or more agreements.
+' The .NET Foundation licenses this file to you under the MIT license.
+' See the LICENSE file in the project root for more information.
+
+Namespace Microsoft.CodeAnalysis.VisualBasic
+
+    Friend NotInheritable Class VisualBasicDeterministicKeyBuilder
+        Inherits DeterministicKeyBuilder
+
+        Friend Shared ReadOnly Instance As New VisualBasicDeterministicKeyBuilder()
+
+        Private Sub New()
+
+        End Sub
+
+        Protected Overrides Sub WriteParseOptionsCore(writer As JsonWriter, parseOptions As ParseOptions)
+            ' This can happen for SyntaxTree that are constructed via the API. 
+            If parseOptions Is Nothing Then
+                Return
+            End If
+
+            Dim basicOptions = TryCast(parseOptions, VisualBasicParseOptions)
+            If basicOptions Is Nothing Then
+                Throw New InvalidOperationException()
+            End If
+
+            MyBase.WriteParseOptionsCore(writer, parseOptions)
+
+            writer.Write("languageVersion", basicOptions.LanguageVersion)
+            writer.Write("specifiedLanguageVersion", basicOptions.SpecifiedLanguageVersion)
+
+            writer.WriteKey("preprocessorSymbols")
+            writer.WriteObjectStart()
+            For Each pair In basicOptions.PreprocessorSymbols.OrderBy(Function(x, y) StringComparer.Ordinal.Compare(x.Key, y.Key))
+                Dim value = pair.Value
+                If value Is Nothing Then
+                    writer.WriteNull(pair.Key)
+                    Continue For
+                End If
+
+                Dim type = value.GetType()
+                If type = GetType(String) Then
+                    writer.Write(pair.Key, CType(value, String))
+                ElseIf type = GetType(Boolean) Then
+                    writer.Write(pair.Key, CType(value, Boolean))
+                ElseIf type = GetType(DateTime) Then
+                    writer.Write(pair.Key, CType(value, DateTime).ToString("G"))
```

Comment:
Consider extracting a `Shared Function ConvertToString(value As Object) As String` helper method.

Reply:
`object.ToString()` though is parmeterless hence I can't force invariant formatting. If there was a way to do this I'd love to hear it cause I don't particularly like this code .